packages/xo-server/src/xo-mixins/pool.js

```diff
@@ -0,0 +1,80 @@
+import { difference, flatten, isEmpty, uniq } from 'lodash'
+
+export default class Pools {
+  constructor(xo) {
+    this._xo = xo
+  }
+
+  async mergeInto({ sources: sourceIds, target, force }) {
+    const { _xo } = this
+    const targetHost = _xo.getObject(target.master)
+    const sources = []
+    const sourcePatches = {}
+    for (const sourceId of sourceIds) {
+      const source = _xo.getObject(sourceId)
+      const sourceHost = _xo.getObject(source.master)
+      if (sourceHost.productBrand !== targetHost.productBrand) {
+        throw new Error(
+          `a ${sourceHost.productBrand} pool cannot be merged into a ${
+            targetHost.productBrand
+          } pool`
+        )
+      }
+      sources.push(source)
+      sourcePatches[sourceId] = sourceHost.patches
+    }
```

Comment:
I think you should identify and comment the main different parts of the algo.

And you should explicitly document the constraints of having to compute the differences prior to do anything in the case of XCP-ng because they can only be updated to the latest state.